src/ngx_http_lua_connection.c

```diff
@@ -0,0 +1,353 @@
+    ngx_reusable_connection(conn, 0);
+
+    /* invalidate connection object */
+    conn->fd = -1;
+    conn->data = NULL;
+    ngx_free_connection(conn);
+}
+
+
+int
+ngx_http_lua_connection_prep(ngx_http_request_t *r, ngx_connection_t *conn,
+    int poll_mask, ngx_msec_t wait_ms, const char **err)
+{
+    ngx_http_lua_ctx_t    *ctx;
+    ngx_http_lua_co_ctx_t *co_ctx;
+    ngx_http_lua_udata_t  *u;
+    if ((ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module)) == NULL) {
+        *err ="no request ctx found";
+        return NGX_ERROR;
+    }
+    if ((co_ctx = ctx->cur_co_ctx) == NULL) {
+        *err ="no co ctx found";
+        return NGX_ERROR;
+    }
+    if ((u = ngx_alloc(sizeof(*u), r->connection->log)) == NULL) {
+        *err ="no memory";
+        return NGX_ERROR;
+    }
+    /* cleanup old events before adding new ones */
+    ngx_http_lua_cleanup_pending_operation(co_ctx);
+    if ((poll_mask & POLLIN) && ngx_add_event(conn->read, NGX_READ_EVENT, NGX_LEVEL_EVENT) != NGX_OK) {
+        ngx_free(u);
+        *err ="unable to add to nginx main loop";
+        return NGX_ERROR;
+    }
+    if ((poll_mask & POLLOUT) && ngx_add_event(conn->write, NGX_WRITE_EVENT, NGX_LEVEL_EVENT) != NGX_OK) {
+        if (poll_mask & POLLIN) {
+            ngx_del_event(conn->read, NGX_READ_EVENT, 0);
+        }
+        ngx_free(u);
+        *err ="unable to add to nginx main loop";
+        return NGX_ERROR;
+    }
+    conn->data = u;
+    u->request = r;
+    u->co_ctx = co_ctx;
+    u->conn = conn;
+    co_ctx->cleanup = (ngx_http_cleanup_pt)&ngx_http_lua_fd_cleanup;
+    co_ctx->data = u;
+
+    if (wait_ms != (ngx_msec_t)-1) {
+        ngx_add_timer(conn->read, wait_ms);
+    }
+
+    /* make sure nginx knows what to do next */
+    if (ctx->entered_content_phase) {
+        r->write_event_handler = ngx_http_lua_content_wev_handler;
+    } else {
+        r->write_event_handler = ngx_http_core_run_phases;
+    }
+
+    return NGX_OK;
+}
+
+
+/* Lua C API bindings */
+
+static int
+ngx_http_lua_connection_new(lua_State *L)
+{
+    ngx_connection_t **pconn;
+    ngx_socket_t       fd = luaL_checkinteger(L, 1);
+    const char        *errmsg;
+    luaL_argcheck(L, fd >= 0, 1, "invalid file descriptor");
+    pconn = lua_newuserdata(L, sizeof(*pconn));
+    *pconn = NULL;
+    luaL_getmetatable(L, NGX_HTTP_LUA_CONNECTION_KEY);
+    lua_setmetatable(L, -2);
+    if (NGX_ERROR == ngx_http_lua_connection_init(pconn, fd, &errmsg)) {
+        return luaL_error(L, errmsg);
+    }
+    (*pconn)->data = NULL;
+    return 1;
+}
+
+
+static ngx_connection_t*
+ngx_http_lua_connection_check(lua_State *L, int idx)
+{
+    ngx_connection_t **pconn = luaL_checkudata(L, idx, NGX_HTTP_LUA_CONNECTION_KEY);
+    luaL_argcheck(L, *pconn != NULL, idx, "ngx_connection_t has been freed");
+    return *pconn;
+}
+
+
+static int
+ngx_http_lua_connection_gc(lua_State *L)
+{
+    ngx_connection_t **pconn = luaL_checkudata(L, 1, NGX_HTTP_LUA_CONNECTION_KEY);
```

Comment:
@daurnimator Sorry, it's not really a C standard thing. It's really an NGINX coding style thing. We should use constant expressions in variable initializations.